pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -36,91 +36,108 @@ import (
 // If the label "__tenant_id__" is already existing, its value is overwritten
 // by the tenant ID and the previous value is exposed through a new label
 // prefixed with "original_". This behaviour is not implemented recursively.
-func NewQueryable(upstream storage.Queryable, byPassWithSingleQuerier bool, maxConcurrency int, logger log.Logger) storage.Queryable {
-	return NewMergeQueryable(defaultTenantLabel, tenantQuerierCallback(upstream), byPassWithSingleQuerier, maxConcurrency, logger)
-}
-
-func tenantQuerierCallback(queryable storage.Queryable) MergeQuerierCallback {
-	return func(ctx context.Context, mint int64, maxt int64) ([]string, []storage.Querier, error) {
-		tenantIDs, err := tenant.TenantIDs(ctx)
-		if err != nil {
-			return nil, nil, err
-		}
-
-		var queriers = make([]storage.Querier, 0, len(tenantIDs))
-		for _, tenantID := range tenantIDs {
-			q, err := queryable.Querier(
-				user.InjectOrgID(ctx, tenantID),
-				mint,
-				maxt,
-			)
+func NewQueryable(upstream storage.Queryable, bypassWithSingleID bool, maxConcurrency int, logger log.Logger) storage.Queryable {
+	callbacks := MergeQuerierCallbacks{
+		Querier: func(mint, maxt int64) (MergeQuerierUpstream, error) {
+			q, err := upstream.Querier(mint, maxt)
 			if err != nil {
-				return nil, nil, err
+				return nil, errors.Wrap(err, "construct querier")
 			}
-			queriers = append(queriers, q)
-		}
 
-		return tenantIDs, queriers, nil
+			return &tenantQuerier{
+				upstream: q,
+			}, nil
+		},
+		IDs: func(ctx context.Context) ([]string, error) {
+			tenantIDs, err := tenant.TenantIDs(ctx)
+			return tenantIDs, err
+		},
 	}
+	return NewMergeQueryable(defaultTenantLabel, callbacks, bypassWithSingleID, maxConcurrency, logger)
+}
+
+type MergeQuerierCallbacks struct {
```

Comment:
I'd suggest to add short doc string to the struct and fields.

Reply:
Done.